### PR TITLE
Handle soft fail for s3keysensor

### PR DESCRIPTION
**Pull request by pankajastro**
if soft_fail is set to true then raise AirflowSkipException. We may need to replicate this to other sensor too

closes: https://github.com/astronomer/astronomer-providers/issues/1149

### CODE REVIEW

File: tests/amazon/aws/sensors/test_s3_sensors.py [hunk 1]
```diff
@@ -243,6 +245,32 @@ def test_s3_key_sensor_with_wildcard_async(self, mock_hook, mock_poke, context):
...
+        [
+            (True, AirflowSkipException),
+            (False, Exception),
+        ],
+    )
+    @mock.patch(f"{MODULE}.S3KeySensorAsync.poke")
+    def test_execute_handle_exception(self, mock_poke, soft_fail, exec):
+        mock_poke.side_effect = Exception()
+        sensor = S3KeySensorAsync(
+            task_id="s3_key_sensor_async", bucket_key="key", bucket_name="bucket", soft_fail=True
```
Comment (pankajastro) [inline tests/amazon/aws/sensors/test_s3_sensors.py:269]:
```suggestion
            task_id="s3_key_sensor_async", bucket_key="key", bucket_name="bucket", soft_fail= soft_fail
```

File: astronomer/providers/amazon/aws/sensors/s3.py [hunk 2]
```diff
@@ -81,20 +81,27 @@ def __init__(
...
+                        poke_interval=self.poke_interval,
+                        soft_fail=self.soft_fail,
+                    ),
+                    method_name="execute_complete",
+                )
+        except Exception as e:
+            if self.soft_fail:
+                raise AirflowSkipException(f"{self.task_id} failed")
+            else:
+                raise e
```
Comment (pankajkoti) [inline astronomer/providers/amazon/aws/sensors/s3.py:104]: Can we try to reduce the scope of the try-except block if this is only to catch the exception in the case of `poke` and not related to `self.defer`?
We could try poke in try-except block. Capture the output of poke. and based on the output of poke, decide to defer or not?

File: astronomer/providers/amazon/aws/sensors/s3.py [hunk 3]
```diff
@@ -81,7 +81,14 @@ def __init__(
 
     def execute(self, context: Context) -> None:
         """Check for a keys in s3 and defers using the trigger"""
-        if not self.poke(context):
+        try:
+            poke = self.poke(context)
+        except Exception as e:
+            if self.soft_fail:
+                raise AirflowSkipException(f"{self.task_id} failed")
```
Comment (pankajkoti) [inline astronomer/providers/amazon/aws/sensors/s3.py:88]:
```suggestion
                raise AirflowSkipException(str(e))
```
To keep this consistent with the output that we use in the triggerer

**Review by pankajkoti — APPROVED** [review]
Can we paste a screenshot of a DAG run when the sensor failed and the task was marked skipped. Looks good to me otherwise. Have a minor comment wrt. to the exception message in the case of poke.

File: tests/amazon/aws/sensors/test_s3_sensors.py [hunk 4]
```diff
@@ -243,6 +245,32 @@ def test_s3_key_sensor_with_wildcard_async(self, mock_hook, mock_poke, context):
...
+        )
+        with pytest.raises(AirflowSkipException):
+            sensor.execute_complete(
+                context={}, event={"status": "error", "message": "mocked error", "soft_fail": True}
+            )
+
+    @pytest.mark.parametrize(
+        "soft_fail,exec",
```
Comment (pankajkoti) [inline tests/amazon/aws/sensors/test_s3_sensors.py:259]:
```suggestion
        "soft_fail,exception",
```
looks like a typo here: `exec`

File: tests/amazon/aws/sensors/test_s3_sensors.py [hunk 5]
```diff
@@ -243,6 +245,32 @@ def test_s3_key_sensor_with_wildcard_async(self, mock_hook, mock_poke, context):
...
+        [
+            (True, AirflowSkipException),
+            (False, Exception),
+        ],
+    )
+    @mock.patch(f"{MODULE}.S3KeySensorAsync.poke")
+    def test_execute_handle_exception(self, mock_poke, soft_fail, exec):
```
Comment (pankajkoti) [inline tests/amazon/aws/sensors/test_s3_sensors.py:266]:
```suggestion
    def test_execute_handle_exception(self, mock_poke, soft_fail, exception):
```

File: tests/amazon/aws/sensors/test_s3_sensors.py [hunk 6]
```diff
@@ -243,6 +245,32 @@ def test_s3_key_sensor_with_wildcard_async(self, mock_hook, mock_poke, context):
...
+        mock_poke.side_effect = Exception()
+        sensor = S3KeySensorAsync(
+            task_id="s3_key_sensor_async", bucket_key="key", bucket_name="bucket", soft_fail=soft_fail
+        )
+        with pytest.raises(exec):
```
Comment (pankajkoti) [inline tests/amazon/aws/sensors/test_s3_sensors.py:271]:
```suggestion
        with pytest.raises(exception):
```

**Comment by pankajastro** [comment]
> Can we paste a screenshot of a DAG run when the sensor failed and the task was marked skipped. Looks good to me otherwise. Have a minor comment wrt. to the exception message in the case of poke.

Screenshot 2023-06-05 at 6 39 17 PM

**Comment by pankajastro** [comment]
Merging this but feel free to raise any concerns I will address them separately

**Pull request merged**